File: web-demo/public/js/linalg.js
// Small dependency-free linear algebra for the calibration fits.
// A faithful port of readsync/src/readsync/webcam/_linalg.py, so the browser
// calibration behaves like the desktop one. Pure functions, unit-tested in Node.

export function solve(matrix, rhs) {
  // Solve a square linear system by Gaussian elimination with partial pivoting.
  const n = matrix.length;
  const a = matrix.map((row, i) => [...row, rhs[i]]);
  for (let col = 0; col < n; col++) {
    let pivot = col;
    for (let r = col + 1; r < n; r++) {
      if (Math.abs(a[r][col]) > Math.abs(a[pivot][col])) pivot = r;
    }
    if (Math.abs(a[pivot][col]) < 1e-12) {
      throw new Error("linear system is singular; vary the targets or add ridge");
    }
    [a[col], a[pivot]] = [a[pivot], a[col]];
    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const factor = a[r][col] / a[col][col];
      for (let k = 0; k <= n; k++) a[r][k] -= factor * a[col][k];
    }
  }
  return a.map((row, i) => row[n] / row[i]);
}

export function ridgeFit(design, target, ridge) {
  // Ridge least-squares coefficients via the regularised normal equations. The
  // penalty is not applied to the constant term at index 0.
  const n = design[0].length;
  const ata = Array.from({ length: n }, () => new Array(n).fill(0));
  const atb = new Array(n).fill(0);
  for (let r = 0; r < design.length; r++) {
    const row = design[r];
    const value = target[r];
    for (let i = 0; i < n; i++) {
      atb[i] += row[i] * value;
      for (let j = 0; j < n; j++) ata[i][j] += row[i] * row[j];
    }
  }
  for (let i = 1; i < n; i++) ata[i][i] += ridge;
  return solve(ata, atb);
}

File: src/readsync/webcam/_linalg.py
"""Small dependency-free linear algebra for the calibration fits.

The webcam calibration solves modest least-squares systems, so a compact, tested
Gaussian-elimination solver is enough and keeps the core free of a numerical
dependency. numpy is present at runtime through the camera stack, but the fits
live here in pure Python so they import and test without it.
"""

from __future__ import annotations

from collections.abc import Sequence

__all__ = ["solve", "ridge_fit"]


def solve(matrix: list[list[float]], rhs: list[float]) -> list[float]:
    """Solve a square linear system by Gaussian elimination with partial pivoting.

    Raises ``ValueError`` if the system is singular, which the callers turn into a
    request to vary the calibration targets or add a ridge penalty.
    """
    n = len(matrix)
    a = [list(row) + [rhs[i]] for i, row in enumerate(matrix)]
    for col in range(n):
        pivot = max(range(col, n), key=lambda r: abs(a[r][col]))
        if abs(a[pivot][col]) < 1e-12:
            raise ValueError("linear system is singular; vary the targets or add ridge")
        a[col], a[pivot] = a[pivot], a[col]
        for r in range(n):
            if r == col:
                continue
            factor = a[r][col] / a[col][col]
            a[r] = [a[r][k] - factor * a[col][k] for k in range(n + 1)]
    return [a[i][n] / a[i][i] for i in range(n)]


def ridge_fit(
    design: Sequence[Sequence[float]], target: Sequence[float], ridge: float
) -> list[float]:
    """Ridge least-squares coefficients via the regularised normal equations.

    The penalty is not applied to the constant term at index 0, so a constant
    offset is never shrunk towards zero. A positive ``ridge`` also keeps the system
    solvable when the design has fewer rows than columns.
    """
    n = len(design[0])
    ata = [[0.0] * n for _ in range(n)]
    atb = [0.0] * n
    for row, value in zip(design, target, strict=True):
        for i in range(n):
            atb[i] += row[i] * value
            for j in range(n):
                ata[i][j] += row[i] * row[j]
    for i in range(1, n):
        ata[i][i] += ridge
    return solve(ata, atb)
